### PR TITLE
Github: obey the rate limit!

### LINKED ISSUE #33: github api: obey speed limit
[`github-api--obey-speed-limit`](https://github.com/iFixit/pulldasher/compare/github-api--obey-speed-limit)

### PR DESCRIPTION
**Pull request by danielbeardsley**
All hail the rate limit!

This adds some smarts to our github usage and will allow requests to
go through unimpeded until we get to the choke point, at which point
we start throttling requests such that we stay *just* under the rate
limit.

Obviously, not knowing the future means we can't be all that smart
about it.

The delay is such that if you have an unlimited number of requests
coming in, we'll stay just below the rate limit. If you reach the
choke point but only have a few more requests, they will be throttled
unnecessarily, but we'll just have to live with that.

This means that now, the rebuild-history option will totally work
without hitting the rate limit. Also, change rebuild-history to
operate on pulls one at a time, instead of thousands in parallel.

Closes #33

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -120,27 +121,33 @@ function refreshAll(all) {
    var githubPulls;
 
    if (all === true) {
+      debug("rebuilding all pulls");
```
Comment (BaseInfinity) [inline app.js]: Not sure if we want to keep these debug statements around but it is kinda nice keeping it around so I'm indifferent about it

File: lib/git-manager.js [hunk 2]
```diff
@@ -6,20 +6,32 @@ var GithubApi = require('github'),
...
+    rateLimit = require('./rate-limit.js');
 
 github.authenticate({
    type: 'oauth',
    token: config.github.token
 });
 
-var get    = Promise.denodeify(github.pullRequests.get);
-var getAll = Promise.denodeify(github.pullRequests.getAll);
+function denodeify(func) {
```
Comment (BaseInfinity) [inline lib/git-manager.js]: I wonder if we should change this function name to show that we are running these functions through rateLimit object

File: lib/git-manager.js [hunk 3]
```diff
@@ -161,17 +168,14 @@ module.exports = {
    },
 
    getPullReviewComments: function(pullNumber) {
-      // "review" comments are the comments on the diff of pull
-      var getPullComments = Promise.denodeify(github.pullRequests.getComments);
```
Comment (BaseInfinity) [inline lib/git-manager.js]: me likey

File: lib/rate-limit.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
+var Promise = require('promise'),
+    debug = require('debug')('pulldasher:rate-limit');
```
Comment (BaseInfinity) [inline lib/rate-limit.js]: there might be a module for this but this is short and concise so im cool with this

**Comment by BaseInfinity** [comment]
left small comments, but otherwise CR 👙 this is schweet

File: app.js (hunk 1 above)
Comment (danielbeardsley) [inline app.js]: The debug module is what we use for debug output (throughout the whole project).

`DEBUG='*' bin/pulldasher` to see all output.

File: lib/git-manager.js (hunk 2 above)
Comment (danielbeardsley) [inline lib/git-manager.js]: Thought about it... but it's too wordy and the function is defined right there. `rateLimitedDenodify` ?

File: lib/rate-limit.js (hunk 4 above)
Comment (danielbeardsley) [inline lib/rate-limit.js]: I don't get it. `debug` *is* a module. We're using it...

**Comment by danielbeardsley** [comment]
Also this is QA 👍  I've run it a bunch while working on it and examined the DB.

**Comment by solflare** [comment]
CR :moon:

File: lib/git-manager.js (hunk 2 above)
Comment (BaseInfinity) [inline lib/git-manager.js]: Lol ya I feel the same way

File: lib/rate-limit.js (hunk 4 above)
Comment (BaseInfinity) [inline lib/rate-limit.js]: Sorry, bad place to leave a comment. I meant a rate-limiting module

Comment (danielbeardsley) [inline lib/rate-limit.js]: The only *rate limiting* part of the code that isn't specific to our use case is the last 25 lines of this file (the rest deals with github headers, choke points, ...).

Never the less, every one I looked at either through away extra calls (i.e. make 10 calls == only 2 get through) and / or had a fixed time interval on creation. We want our interval to be flexible and respond to the amount of time / requests remaining.
